sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1530,6 +1650,18 @@ class Analyzer(
   object GlobalAggregates extends Rule[LogicalPlan] {
     def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperators {
       case Project(projectList, child) if containsAggregates(projectList) =>
+
+        def traversePlanAndCheck(plan: LogicalPlan): Unit =
```

Comment:
Got a similar comment from @viirya , I will try to refactor error handling a bit.

Reply:
I can't move this check to `CTESubstitution` because the time when it runs aggregate functions haven't been resolved yet and `containsAggregates()` doesn't work.

btw we need this in this case:
```
-- recursive reference is not allowed in an aggregate (made from project) in a recursive term
WITH RECURSIVE r AS (
  VALUES (1L) AS T(data)
  UNION ALL
  SELECT SUM(data) FROM r WHERE data < 10
)
SELECT * FROM r;
```
when there is a projection only in the beginning and then that projection is then transformed to an aggregate in 'GlobalAggregates', that's why I thought this would be a good place to do the test.

Do you think moving this one check to a dedicated "checker" rule is better?

Reply:
yes, I'd move all the safety checks to a dedicated rule

Reply:
Ok then I will try to move all the check into that rule.